jni/lib/src/jni_object.dart

```diff
@@ -7,178 +7,543 @@ import 'dart:ffi';
+      result = ctor(ref) as T;
+      break;
+    case _VoidType:
+      finalCallType =
+          _getCallType(callType, JniType.voidType, {JniType.voidType});
+      f(finalCallType);
+      result = null as T;
+      break;
+    case dynamic:
+      result = f(callType ?? JniType.voidType) as T;
+      break;
+    default:
+      throw UnsupportedError('Unknown type $T');
+  }
+  return result;
+}
+
+T _callMethod<T>(
+    int? callType, List<dynamic> args, Function(int, Pointer<JValue>) f) {
+  final jArgs = JValueArgs(args);
+  final result = _callOrGet<T>(callType, (ct) => f(ct, jArgs.values));
+  calloc.free(jArgs.values);
+  jArgs.dispose();
+  if (result == 0 || result == 0.0 || result == null) {
+    _env.checkException();
+  }
+  return result;
+}
+
+T _getField<T>(int? callType, Function(int) f) {
+  final result = _callOrGet<T>(callType, f);
+  _env.checkException();
+  return result;
+}
+
+/// A high-level wrapper for JNI global object reference.
 ///
-/// It's valid only in the thread it was created.
-/// When passing to code that might run in a different thread (eg: a callback),
-/// consider obtaining a global reference and reconstructing the object.
-class JniObject {
-  JClass _cls;
-  final JObject _obj;
-  final Pointer<JniEnv> _env;
+/// This is the base class for classes generated by `jnigen`.
+class JniObject extends JniReference implements Finalizable {
+  /// Finalizer for this class
+  static final _finalizer = NativeFinalizer(_env.ref.DeleteGlobalRef);
+
+  /// Construct a new [JniObject] with [reference] as its underlying reference.
+  JniObject.fromRef(this.reference) {
+    _finalizer.attach(this, reference, detach: this);
+  }
+
+  /// Pass the underlying reference into [function] and return the result.
+  ///
+  /// Result of [function] takes ownership of [reference], and this object is
+  /// marked as deleted. [function] can be constructor of a subclass of JniClass
+  /// in which case this function will be useful to cast an expression.
+  T into<T extends JniObject>(T Function(JObject) function) {
+    _assertNotDeleted();
+    _finalizer.detach(this);
+    _deleted = true;
+    return function(reference);
+  }
+
+  /// Returns whether this object is deleted.
+  @override
+  bool get isDeleted => _deleted;
+
+  /// Check whether the underlying JNI reference is `null`.
+  bool get isNull => reference == nullptr;
+
+  /// The underlying JNI global reference of this object.
+  final JObject reference;
+
   bool _deleted = false;
-  JniObject.of(this._env, this._obj, this._cls);
 
-  @pragma('vm:prefer-inline')
-  void _checkDeleted() {
+  /// Reference to class of the reference obtained using `GetObjectClass`. It's
+  /// populated only if any methods called on this object require it.
+  ///
+  /// This is nullable field instead of a late field, because we need to check
+  /// if it's populated and delete if so.
+  JObject? _classRef;
+
+  JObject get _class {
```

Comment:
Actually most reflective methods require class reference. Eg callMethodByName() or getFieldId().

So `obj.callMethodByName()` becomes either
`obj.callMethod(obj.getClass().getMethod(...)` or obj requires to have a late initialized `JniClass` member instead.

Reply:
> `obj.callMethod(obj.getClass().getMethod(name, sig), args)`

Ah that would be bad. I had only seen `getFieldID` which doesn't need an object, only the class.

> obj requires to have a late initialized `JniClass` member instead.

Yes, that is a better solution. (Since `JniObject` and `JniClass` are in the same file, JniObject should have access to JniClass internal _reference.)

Reply:
Last question. Having late initialized JniClass means it won't be deleted in eg: an arena. So now we have to make it nullable, check if initialized and delete it in JniObject's delete() method. Doesn't that mean we are almost back to square one?

Reply:
> So now we have to make it nullable, check if initialized and delete it in JniObject's delete() method.

Correct.

> Doesn't that mean we are almost back to square one?

Yes and no. Yes, we have a check. No, if we forgot to delete it, the NativeFinalizer would delete it.

I think that is the cleanest solution. (I'll try to think some more about what kind of patterns would be nice with kind of coupled native resources.) This is really an artifact of the JNI API where you have to pass the object _and_ the class for those methods. And we want to provide something nicer for our users.